python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -0,0 +1,370 @@
+# pylint: disable=anomalous-backslash-in-string
+import json
+from collections import defaultdict
+from typing import TYPE_CHECKING, Dict, List, Mapping, Optional, Sequence, Set, Tuple
+
+import toposort
+
+from dagster._annotations import experimental
+from dagster._core.storage.pipeline_run import IN_PROGRESS_RUN_STATUSES, RunsFilter
+
+from .asset_selection import AssetSelection
+from .events import AssetKey
+from .run_request import RunRequest
+from .sensor_definition import DefaultSensorStatus, MultiAssetSensorDefinition
+from .utils import check_valid_name
+
+if TYPE_CHECKING:
+    from dagster._core.definitions import AssetsDefinition, SourceAsset
+    from dagster._core.storage.event_log.base import EventLogRecord
+
+
+def _get_upstream_mapping(
+    selection,
+    assets,
+    source_assets,
+) -> Mapping[AssetKey, Set[AssetKey]]:
+    """Computes a mapping of assets in self._selection to their parents in the asset graph"""
+    upstream = defaultdict(set)
+    selection_resolved = list(selection.resolve([*assets, *source_assets]))
+    for a in selection_resolved:
+        a_parents = list(
+            AssetSelection.keys(a).upstream(depth=1).resolve([*assets, *source_assets])
+        )
+        # filter out a because upstream() includes the assets in the original AssetSelection
+        upstream[a] = {p for p in a_parents if p != a}
+    return upstream
+
+
+def _get_parent_updates(
+    context,
+    current_asset: AssetKey,
+    parent_assets: Set[AssetKey],
+    cursor_timestamp: float,
+    will_materialize_list: Sequence[AssetKey],
+    wait_for_in_progress_runs: bool,
+) -> Mapping[AssetKey, Tuple[bool, float]]:
+    """The bulk of the logic in the sensor is in this function. At the end of the function we return a
+    dictionary that maps each asset to a Tuple. The Tuple contains a boolean, indicating if the asset
+    has materialized or will materialize, and a float representing the timestamp the parent asset
+    would update the cursor too if it is the most recent materialization of a parent asset. In some cases
+    we set the timestamp to 0.0 so that the timestamps of other parent materializations will take precedent.
+
+    Here's how we get there:
+
+    We want to get the materialization information for all of the parents of an asset to determine
+    if we want to materialize the asset in this sensor tick. We also need determine the new cursor
+    value for the asset so that we don't process the same materialization events for the parent
+    assets again.
+
+    We iterate through each parent of the asset and determine its materialization info. The parent
+    asset's materialization status can be one of three options:
+    1. The parent has materialized since the last time the child was materialized (determined by comparing
+        the timestamp of the parent materialization to the cursor_timestamp).
+    2. The parent will be materialized as part of the materialization that will be kicked off by the
```

Comment:
would it be accurate to say "The parent is already slated to be materialized (i.e. included in `will_materialize_list`"?

this tripped me up a little bit on first read, because I first interpreted it (I think erroneously?) as that this function was both finding out info about parents and also making decisions about would be materialized.

Reply:
yeah that's clearer